src/devices/vdml_link.c

```diff
@@ -0,0 +1,234 @@
+    // claim_port_i(port - 1, E_DEVICE_RADIO);
+	if (!VALIDATE_PORT_NO(port - 1)) {
+		errno = EINVAL;
+		return PROS_ERR;
+	}
+    // by default, the vexlink radio in the lower port should be by default E_DEVICE_NONE
+    // or, the only radio on the brain
+    v5_device_e_t plugged_device = registry_get_plugged_type(port - 1);
+    if(plugged_device != E_DEVICE_NONE && plugged_device != E_DEVICE_RADIO) {
+        return PROS_ERR;
+    }
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	if (!port_mutex_take(port - 1)) {
+		errno = EACCES;
+		return PROS_ERR;
+	}
+    vexDeviceGenericRadioConnection(device->device_info, link_id, type, false);
+    return_port(port - 1, 1);
+}
+
+uint32_t link_init_override(uint8_t port, char* link_id, link_type_e_t type) {
+    // not used because it's a generic serial device 
+    // claim_port_i(port - 1, E_DEVICE_RADIO);
+	if (!VALIDATE_PORT_NO(port - 1)) {
+		errno = EINVAL;
+		return PROS_ERR;
+	}
+    // by default, the vexlink radio in the lower port should be by default E_DEVICE_NONE
+    // or, the only radio on the brain
+    v5_device_e_t plugged_device = registry_get_plugged_type(port - 1);
+    if(plugged_device != E_DEVICE_NONE && plugged_device != E_DEVICE_RADIO) {
+        return PROS_ERR;
+    }
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	if (!port_mutex_take(port - 1)) {
+		errno = EACCES;
+		return PROS_ERR;
+	}
+    vexDeviceGenericRadioConnection(device->device_info, link_id, type, true);
+    return_port(port - 1, 1);
+}
+
+bool link_connected(uint8_t port) {
+    claim_port(port - 1, E_DEVICE_SERIAL, false);
+    bool rtv = vexDeviceGenericRadioLinkStatus(device->device_info); 
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_raw_receivable_size(uint8_t port) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    uint32_t rtv = vexDeviceGenericRadioReceiveAvail(device->device_info); 
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_raw_transmittable_size(uint8_t port) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    uint32_t rtv = vexDeviceGenericRadioWriteFree(device->device_info);
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_transmit_raw(uint8_t port, void* data, uint16_t data_size) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    if(!vexDeviceGenericRadioLinkStatus(device->device_info)) {
+        errno = ENXIO;
+        return_port(port - 1, PROS_ERR);
+    }
+    if(data_size > vexDeviceGenericRadioWriteFree(device->device_info)) {
+        errno = EBUSY;
+        return_port(port - 1, 0);
+    }
+    if(data == NULL) {
+        errno = EINVAL;
+        return_port(port - 1, PROS_ERR);
+    }
+    vexDeviceGenericRadioTransmit(device->device_info, (uint8_t*)data, data_size);
+    return_port(port - 1, 1);
+}
+
+uint32_t link_receive_raw(uint8_t port, void* dest, uint16_t data_size) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL); 
+    if(data_size > vexDeviceGenericRadioReceiveAvail(device->device_info)) {
```

Comment:
> If you do: // send 256 bytes of data // send 512 bytes of data wait // receive 256 bytes of data // receive 512 bytes of data
> 
> what happens?

This should ignore both 512 commands, as there wouldn't be enough size in the TX buffer to transfer or enough bytes to read in the RX buffer.

Reply:
What about: 
// send 256 bytes of data
wait a bit
// send 512 bytes of data
wait
// receive 256 bytes of data
wait a bit
// receive 512 bytes of data

I assume it still truncates the last 512 bytes and only returns 256? Or does it drop the entire message?

Reply:
It drops the entire message (our implementation does at the very least), but the vexDevice API truncates it

Reply:
For receives I still kind of think we should provide a way of doing partial reads of not the full data size. It probably should be a separate method, since this current function will make sending/receiving a struct directly easy. 

My main idea is that they can throttle bandwidth on the sender, and read in the receiving buffer on like a 1/5 ms loop. That way you can stream data without having to send an ack to the original sender to keep sending data. Though tbh, idk how realistic that scenario is. I don't know why they'd send a significant amount of data over the radio.

Reply:
> That way you can stream data without having to send an ack to the original sender to keep sending data.

We currently aren't doing this though, but I understand where you're coming from with the partial RX data. However, unless they're using the raw methods, the data is packeted so allowing for partial data may lead for partial packets on the rx end as well.